### PR TITLE
Task 11456 logging system

### PR DESCRIPTION
**Pull request by jakubzimnol**

### CODE REVIEW

File: backend/app/models.py [hunk 1]
```diff
@@ -142,3 +150,13 @@ class Image(db.Model):
 
     def __repr__(self):
         return '<Image %r>' % self.name
+
+
+class RevokedTokenModel(db.Model):
+    id = db.Column(db.Integer, primary_key=True)
+    jti = db.Column(db.String(120))
```
Comment (mlugowska) [inline backend/app/models.py]: What is `jti`?

File: backend/app/resources.py [hunk 2]
```diff
@@ -33,25 +37,27 @@ class ItemsList(Resource):
...
         return marshal(Item.query.all(), item_marshaller)
 
+    @jwt_required
     def post(self):
         args = creating_item_parser.parse_args()
         item = Repository.create_and_add(Item, args)
         if not item:
-            return 404
-        db.session.commit()
+            return {'message': 'Something went wrong'}, 500
```
Comment (mlugowska) [inline backend/app/resources.py]: Aaa, backend should never return 500! And `Something went wrong` is probably the worst message, something like `Something is not okay, but I won't tell you what you need to guess, let's play a game, hehe`

File: backend/app/resources.py [hunk 3]
```diff
@@ -64,24 +70,27 @@ class CategoryList(Resource):
...
         return marshal(Category.query.all(), category_marshaller)
 
+    @jwt_required
     def post(self):
         args = creating_category_parser.parse_args()
-        category = Repository.create_and_add(Item, args)
+        category = Repository.create_and_add(Category, args)
         if not category:
-            return 404
+            return {'message': 'Something went wrong'}, 500
```
Comment (mlugowska) [inline backend/app/resources.py]: same as above

File: backend/app/resources.py [hunk 4]
```diff
@@ -94,9 +103,86 @@ class SubcategoryList(Resource):
...
         return marshal(Subcategory.query.all(), subcategory_marshaller)
 
+    @jwt_required
     def post(self):
         args = creating_subcategory_parser.parse_args()
-        subcategory = Repository.create_and_add(args)
+        subcategory = Repository.create_and_add(Subcategory, args)
         if not subcategory:
-            return 404
+            return {'message': 'Something went wrong'}, 500
```
Comment (mlugowska) [inline backend/app/resources.py]: same as above

File: backend/app/resources.py [hunk 5]
```diff
@@ -94,9 +103,86 @@ class SubcategoryList(Resource):
...
+
+
+class UserRegistration(Resource):
+    def post(self):
+        args = user_parser.parse_args()
+
+        try:
+            new_user = User(username=args['username'])
+            db.session.add(new_user)
+            new_user.password_hash = args['password']
```
Comment (mlugowska) [inline backend/app/resources.py]: This `try/except` block does far too much. Split it into smaller methods or something.

File: backend/app/resources.py [hunk 6]
```diff
@@ -94,9 +103,86 @@ class SubcategoryList(Resource):
...
+        if current_user.check_password(args['password']):
+            access_token = create_access_token(identity=args['username'])
+            refresh_token = create_refresh_token(identity=args['username'])
+            return {
+                       'message': 'Logged in as {}'.format(current_user.username),
+                       'access_token': access_token,
+                       'refresh_token': refresh_token
+                   }, 201
+        else:
+            return {'message': 'Wrong credentials'}
```
Comment (mlugowska) [inline backend/app/resources.py]: status code?

Comment (mlugowska) [reply backend/app/resources.py]: You don't need to user `else` if in `if` you have an `return`.

File: backend/app/resources.py [hunk 7]
```diff
@@ -94,9 +103,86 @@ class SubcategoryList(Resource):
...
+
+
+class UserLogoutAccess(Resource):
+    @jwt_required
+    def post(self):
+        jti = get_raw_jwt()['jti']
+        try:
+            revoked_token = RevokedTokenModel(jti=jti)
+            db.session.add(revoked_token)
```
Comment (mlugowska) [inline backend/app/resources.py]: I believe that `db.session.add` and `commit` you have as `add_to_database` method. Why don't use it? :D

File: backend/app/resources.py [hunk 8]
```diff
@@ -94,9 +103,86 @@ class SubcategoryList(Resource):
...
+            return {'message': 'Something went wrong'}, 500
+
+
+class UserLogoutRefresh(Resource):
+    @jwt_refresh_token_required
+    def post(self):
+        jti = get_raw_jwt()['jti']
+        try:
+            revoked_token = RevokedTokenModel(jti=jti)
+            db.session.add(revoked_token)
```
Comment (mlugowska) [inline backend/app/resources.py]: same as above

**Review by mlugowska — CHANGES_REQUESTED** [review]

File: backend/app/models.py (hunk 1 above)
Comment (jakubzimnol) [reply backend/app/models.py]: jti = JWT ID 
this Class stores revoked tokens

File: backend/init_app.py [hunk 9]
```diff
@@ -29,6 +47,12 @@
 api.add_resource(Categories, '/api/category/<category_id>')
 api.add_resource(SubcategoryList, '/api/subcategory')
 api.add_resource(Subcategories, '/api/subcategory/<subcategory_id>')
+api.add_resource(UserRegistration, '/api/registration')
+api.add_resource(UserLogin, '/api/login')
+api.add_resource(UserLogoutAccess, '/api/logout/access')
```
Comment (JerzySpendel) [inline backend/init_app.py:61]: `/api/logout/access` -  what a strange URL :D What is it supposed to mean? If it's to logout a user, then POST on `/api/logout/` is enough

File: backend/init_app.py [hunk 10]
```diff
@@ -29,6 +47,12 @@
 api.add_resource(Categories, '/api/category/<category_id>')
 api.add_resource(SubcategoryList, '/api/subcategory')
 api.add_resource(Subcategories, '/api/subcategory/<subcategory_id>')
```
Comment (JerzySpendel) [inline backend/init_app.py]: Sorry, this does not belong to this PR but I haven't seen it any earlier. I guess from name that subcategory is always under some category, so le'ts mirror it in URL schema! :)

```suggestion
api.add_resource(Subcategories, '/api/categories/<category_id>/subcategory/<subcategory_id>')
```
BTW Resources URL have to be in plural form.
http://www.andrewhavens.com/posts/20/beginners-guide-to-creating-a-rest-api/ (I have to read it too... :D)

File: backend/app/models.py (hunk 1 above)
Comment (JerzySpendel) [reply backend/app/models.py]: No need to short a shortcut :) Call it `jwt_id`

File: backend/app/models.py [hunk 11]
```diff
@@ -142,3 +150,13 @@ class Image(db.Model):
...
+
+    @classmethod
+    def is_jti_blacklisted(cls, jti):
+        query = cls.query.filter_by(jti=jti).first()
+        return bool(query)
```
Comment (JerzySpendel) [inline backend/app/models.py]: Looks like you only want to determine if a given JWT matching some conditions exists, so why you order DB to download this JWT? :)

Use `.exists()` method, I haven't checked if there is such a method, but I believe there is.

**Review by JerzySpendel — CHANGES_REQUESTED** [review]

File: backend/init_app.py (hunk 9 above)
Comment (jakubzimnol) [reply backend/init_app.py:61]: I have to add refresh token and access token to backlist. But Refresh tokens cannot access an endpoint that is protected with jwt_required() and access tokens cannot access and endpoint that is protected with jwt_refresh_token_required() so I need 2 endpoints one for logout acccess token and one for logout refresh token.

File: backend/app/exceptions.py [hunk 12]
```diff
@@ -0,0 +1,16 @@
+
```
Comment (wojciech-wolny) [inline backend/app/exceptions.py:4]: Why empty line?

File: backend/app/exceptions.py [hunk 13]
```diff
@@ -0,0 +1,16 @@
...
+class MyBaseException(Exception):
```
Comment (wojciech-wolny) [inline backend/app/exceptions.py]: Why `My` just `BaseException` would be fine or `APIBaseException`

File: backend/app/models.py [hunk 14]
```diff
@@ -1,15 +1,17 @@
...
+from sqlalchemy import exists
 from sqlalchemy.ext.declarative import declarative_base
 from sqlalchemy.ext.hybrid import hybrid_property
 from werkzeug.security import generate_password_hash, check_password_hash
 
 from init_app import db
-from abc import ABC, abstractmethod
 
 
 class ObjectABC(ABC):
```
Comment (wojciech-wolny) [inline backend/app/models.py]: Object ABC says nothing about AbstractClass, what this abstract class do? Or what kind of classes will use this ABC

File: backend/app/parsers.py [hunk 15]
```diff
@@ -17,23 +17,57 @@ def check_subcategory(value):
     return value
 
 
+def check_unique_item(value):
+    if db.session.query(exists().where(Item.name == value)).scalar():
```
Comment (wojciech-wolny) [inline backend/app/parsers.py]: Do you see what is common in those 4 functions?

File: backend/app/resources.py [hunk 16]
```diff
@@ -1,25 +1,41 @@
...
 from app.repositories import Repository
 from init_app import db
 
 
+def create_tokens(username):
+    try:
+        access_token = create_access_token(identity=username)
+        refresh_token = create_refresh_token(identity=username)
+        return access_token, refresh_token
+    except Exception:
```
Comment (wojciech-wolny) [inline backend/app/resources.py]: don't except Exception, is similar to empty except... What kind of error do you want to catch?

File: backend/app/resources.py [hunk 17]
```diff
@@ -94,9 +116,82 @@ class SubcategoryList(Resource):
...
+        except CreateTokenException as exc:
+            return exc.message, exc.status_code
+
+
+class UserLogin(Resource):
+    def post(self):
+        args = login_user_parser.parse_args()
+        current_user = User.query.filter_by(username=args['username']).first_or_404()
+        if current_user.check_password(args['password']):
+            try:
```
Comment (wojciech-wolny) [inline backend/app/resources.py]: this the `if` block is quite big don't you think?

**Review by wojciech-wolny — CHANGES_REQUESTED** [review]

File: backend/app/parsers.py [hunk 18]
```diff
@@ -1,39 +1,79 @@
...
-from app.models import Category, Subcategory
+from app.models import Category, Subcategory, Item, User
 from init_app import db
 
 
-def check_category(value):
-    if not db.session.query(exists().where(Category.name == value)).scalar():
-        raise ValueError("Value is not Category instance")
+def object_exist(model, value):
+    return db.session.query(exists().where(model.name == value)).scalar()
```
Comment (wojciech-wolny) [inline backend/app/parsers.py]: You think you want to check object existence by name every time?

File: backend/app/repositories.py [hunk 19]
```diff
@@ -9,15 +10,17 @@ def add_to_database(item):
         try:
             db.session.add(item)
             db.session.commit()
-            return True
-        except exc.IntegrityError:
+        except exc.IntegrityError as e:
             db.session.rollback()
-            return False
+            raise IntegrityException()
```
Comment (wojciech-wolny) [inline backend/app/repositories.py:15]: just raise should be fine

**Review by wojciech-wolny — CHANGES_REQUESTED** [review]

**Review by wojciech-wolny — APPROVED** [review]

File: backend/app/exceptions.py [hunk 20]
```diff
@@ -0,0 +1,11 @@
+class ApiBaseException(Exception):
+    def __init__(self, status_code=400, message='undefined exception'):
+        self.status_code = status_code
+        self.message = message
+
+    def add_message(self, msg):
```
Comment (JerzySpendel) [inline backend/app/exceptions.py]: Looks like you don't use that method anywhere?

File: backend/app/parsers.py [hunk 21]
```diff
@@ -1,39 +1,77 @@
...
-def check_subcategory(value):
-    if not db.session.query(exists().where(Subcategory.name == value)).scalar():
-        raise ValueError("Value is not Subcategory instance")
+def check_unique_object(model, value, name='name'):
+    if object_exist(model, value, name):
+        raise ValueError(f"{value} already exist")
     return value
 
 
+def check_category(value):
```
Comment (JerzySpendel) [inline backend/app/parsers.py]: Try to create this function with `partial` from `functools` package :) It should be in form as below:
`check_category = partial(...)`

Try to do this :)

File: backend/app/resources.py [hunk 22]
```diff
@@ -35,23 +46,25 @@ def get(self):
 
     def post(self):
         args = creating_item_parser.parse_args()
-        item = Repository.create_and_add(Item, args)
-        if not item:
-            return 404
-        db.session.commit()
-        return marshal(item, item_marshaller), 201
+        try:
```
Comment (JerzySpendel) [inline backend/app/resources.py]: Once you implement error handler this `try..except` should no longer be necessary

**Review by JerzySpendel — CHANGES_REQUESTED** [review]

File: backend/app/exceptions.py [hunk 23]
```diff
@@ -0,0 +1,13 @@
+class ApiBaseException(Exception):
+    message = None
+    status_code = 400
+
+    def __init_subclass__(cls, **kwargs):
+        assert cls.message
+
+    def to_dict(self):
```
Comment (JerzySpendel) [inline backend/app/exceptions.py:8]: I think this method will not return dictionaries.

File: backend/app/models.py [hunk 24]
```diff
@@ -142,3 +151,12 @@ class Image(db.Model):
...
     def __repr__(self):
         return '<Image %r>' % self.name
+
+
+class RevokedTokenModel(db.Model):
+    id = db.Column(db.Integer, primary_key=True)
+    jwt_id = db.Column(db.String(120))
+
+    @classmethod
+    def is_jti_blacklisted(cls, jwt_id):
```
Comment (JerzySpendel) [inline backend/app/models.py]: By context (class where this is defined) it is obvious that this method is about JWT, so you could call it `is_blacklisted`

File: backend/app/parsers.py [hunk 25]
```diff
@@ -1,39 +1,70 @@
...
+def check_object(model, value, attribute='name'):
+    if not object_exist(model, value, attribute):
+        raise ValueError(f"{value} is not class instance")
     return value
 
 
-def check_subcategory(value):
-    if not db.session.query(exists().where(Subcategory.name == value)).scalar():
-        raise ValueError("Value is not Subcategory instance")
+def check_unique_object(model, value, attribute='name'):
```
Comment (JerzySpendel) [inline backend/app/parsers.py]: I know I suggested making partial function from this function, it was excercise for you. Is this function really needed? Do you make unique constraint on the models?

**Review by JerzySpendel — CHANGES_REQUESTED** [review]

File: backend/app/parsers.py (hunk 25 above)
Comment (jakubzimnol) [reply backend/app/parsers.py]: Yes, I do

File: backend/app/models.py [hunk 26]
```diff
@@ -1,15 +1,17 @@
...
 from sqlalchemy.ext.hybrid import hybrid_property
 from werkzeug.security import generate_password_hash, check_password_hash
 
 from init_app import db
-from abc import ABC, abstractmethod
 
 
-class ObjectABC(ABC):
+class AbstractUpdater(ABC):
     name = NotImplemented
```
Comment (JerzySpendel) [inline backend/app/models.py]:
```suggestion
    name = None
```

File: backend/app/models.py [hunk 27]
```diff
@@ -142,3 +151,12 @@ class Image(db.Model):
 
     def __repr__(self):
         return '<Image %r>' % self.name
+
+
+class RevokedTokenModel(db.Model):
+    id = db.Column(db.Integer, primary_key=True)
+    jwt_id = db.Column(db.String(120))
```
Comment (JerzySpendel) [inline backend/app/models.py]:
```suggestion
    jti = db.Column(db.String(120))
```

File: backend/app/parsers.py (hunk 25 above)
Comment (JerzySpendel) [inline backend/app/parsers.py]:
```suggestion
def check_unique_object(model, value, attribute='column'):
```

**Review by JerzySpendel — APPROVED** [review]
